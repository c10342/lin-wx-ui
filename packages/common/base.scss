@import "../../src/style/var.scss";

.lin-border-top {
  position: relative;
  &::before {
    top: 0;
    position: absolute;
    left: 0;
    right: 0;
    content: "";
    height: 1px;
    background-color: $gray-3;
  }
  &-hairline {
    &::before {
      transform: scaleY(0.5);
    }
  }
}

.lin-border-bottom {
  position: relative;
  &::after {
    bottom: 0;
    position: absolute;
    left: 0;
    right: 0;
    content: "";
    height: 1px;
    background-color: $gray-3;
  }
  &-hairline {
    &::after {
      transform: scaleY(0.5);
    }
  }
}

.lin-border-surround {
  position: relative;
  &::after {
    position: absolute;
    box-sizing: border-box;
    content: " ";
    pointer-events: none;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    border: 1px solid $gray-3;
  }
  &-hairline {
    transform: scale(0.5);
    top: -50%;
    right: -50%;
    bottom: -50%;
    left: -50%;
  }
}

.one-line {
  overflow: hidden;
  -ms-text-overflow: ellipsis;
  text-overflow: ellipsis;
  white-space: nowrap;
}
