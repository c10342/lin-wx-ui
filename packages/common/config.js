export const publicPath = 'https://wxui.linjiafu.top';

export const staticPath = `${publicPath}/static`;

export const imagePath = `${staticPath}/images`;

export const jsonPath = `${staticPath}/json`;

export const emptyCompImages = {
  default: `${imagePath}/default.png`,
  error: `${imagePath}/error.png`,
  network: `${imagePath}/network.png`,
  search: `${imagePath}/search.png`
};

export const areaJsonPath = `${jsonPath}/area.json`;
