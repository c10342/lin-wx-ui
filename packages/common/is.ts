const toString = Object.prototype.toString;

export function isFunction(data: any): data is Function {
  return toString.call(data) === "[object Function]";
}

export function isString(data: any): data is string {
  return toString.call(data) === "[object String]";
}

export function isNull(data: any): data is null {
  return toString.call(data) === "[object Null]";
}

export function isUndefined(data: any): data is undefined {
  return toString.call(data) === "[object Undefined]";
}

export function isUndef(data: any): data is null | undefined {
  return isUndefined(data) || isNull(data);
}

export function isObject(obj: any): obj is Object {
  const type = typeof obj;
  return obj !== null && (type === "object" || type === "function");
}
