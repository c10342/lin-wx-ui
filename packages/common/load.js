#!/usr/bin/env node

const fs = require('fs');

const path = require('path');

const root = process.cwd();

// lin-wx-ui配置文件
const config = require(path.join(root, './lin-wx-ui.config.json'));

// 项目配置文件
const projectConfigSrc = path.join(root, './project.config.json');

// 组件库路径
const lib = path.join(root, config.lib);

const appJson = require(path.join(root, './app.json'));

// 获取注册的页面
const pages = appJson.pages || [];

// 获取全局注册的页面
const globalComponents = appJson.usingComponents || {};

// 记录被使用到的lin-wx-ui组件
const components = {};

// 记录已经扫描过的组件，防止循环引用
const scaleComponent = [];

// 处理全局注册的组件
function handleGlobalComponent() {
  Object.keys(globalComponents).forEach((key) => {
    // // 获取全局组件的完整路径
    const compSrc = path.join(root, globalComponents[key]);
    if (scaleComponent.findIndex((item) => item === compSrc) === -1) {
      // 判断该组件是否属于lin-wx-ui的
      if (compSrc.startsWith(lib)) {
        // 以key-value的形式存储
        const objKey = compSrc.replace(lib, '');
        if (!(objKey in components)) {
          // 记录相对路径
          components[objKey] = path.relative(root, path.dirname(compSrc));
          handlePageComponent(root, globalComponents[key]);
        }
      } else {
        handlePageComponent(root, globalComponents[key]);
      }
      scaleComponent.push(compSrc);
    }
  });
}

/**
 *
 * @param {*} currentDirname 当前路径
 * @param {*} page 相对路径
 */
function handlePageComponent(currentDirname, page) {
  page = path.normalize(page);
  // 获取页面/组件的完整路径
  const pageSrc = path.join(currentDirname, page);
  // 获取页面/组件所在的目录
  const pageDirname = path.dirname(pageSrc);
  // 获取文件名
  const name = path.basename(pageSrc);
  // 获取目录下的json的文件路径
  const pageJsonSrc = path.join(pageDirname, `${name}.json`);
  // 判断json文件是否存在
  const status = fs.existsSync(pageJsonSrc);
  if (status) {
    // 获取json文件的内容
    const pageJsonStr = fs.readFileSync(pageJsonSrc, 'utf-8');
    // 获取json文件使用的组件
    const usingComponents = JSON.parse(pageJsonStr).usingComponents || {};
    Object.keys(usingComponents).forEach((key) => {
      const compVal = usingComponents[key];
      // 判断是否为绝对路径
      const isAbsolute = path.isAbsolute(compVal);
      let compSrc = '';
      if (isAbsolute) {
        compSrc = path.join(currentDirname, compVal);
      } else {
        compSrc = path.join(pageDirname, compVal);
      }
      if (scaleComponent.findIndex((item) => item === compSrc) === -1) {
        if (compSrc.startsWith(lib)) {
          // 判断该组件是否属于lin-wx-ui的
          const objKey = compSrc.replace(lib, '');
          if (!(objKey in components)) {
            // 组件还没收集
            components[objKey] = path.relative(root, path.dirname(compSrc));
            handlePageComponent(path.dirname(compSrc), `..${objKey}`);
          }
        } else {
          handlePageComponent(path.dirname(compSrc), compSrc);
        }
        scaleComponent.push(compSrc);
      }
    });
  }
}

handleGlobalComponent();

pages.forEach((page) => {
  handlePageComponent(root, page);
});

// 求交集，就是找出那些是没有使用的组件
const ignoreArr = [];

const dirs = fs.readdirSync(lib);

const componentsKeys = Object.keys(components);

const ignoreComponent = [];

const reg = /^[A-Z]/;
dirs.forEach((dir) => {
  const index = componentsKeys.findIndex((key) => key.includes(dir));
  if (index === -1 && reg.test(dir)) {
    // 没有使用到的组件
    const ps = path.relative(root, path.join(lib, dir));
    ignoreComponent.push(ps);
  }
});

ignoreComponent.forEach((key) => {
  ignoreArr.push({
    type: 'folder',
    value: key.replace(/\\/g, '/'),
  });
});

const projectConfig = JSON.parse(fs.readFileSync(projectConfigSrc)) || {};

const relativeLib = path.relative(root, lib);

if (!projectConfig.packOptions) {
  projectConfig.packOptions = {};
}

let ignore = projectConfig.packOptions.ignore || [];

// 先过滤掉上次写入进去的
ignore = ignore.filter((src) => {
  const value = path.normalize(src.value);
  return !value.startsWith(relativeLib);
});

ignore.push(...ignoreArr);

if (ignore.findIndex((item) => item.value === 'lin-wx-ui.config.json') === -1) {
  ignore.push({
    type: 'file',
    value: 'lin-wx-ui.config.json',
  });
}

projectConfig.packOptions.ignore = ignore;

fs.writeFileSync(projectConfigSrc, JSON.stringify(projectConfig, null, 2));

/* eslint-disable-next-line */
console.log('执行完毕');
