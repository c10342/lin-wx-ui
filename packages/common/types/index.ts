import { Weapp } from "./weapp";

type RecordToAny<T> = { [K in keyof T]: any };

export type CombinedComponentInstance<Data, Props, Methods> = Methods &
  WechatMiniprogram.Component.TrivialInstance &
  Weapp.FormField & {
    data: Data & RecordToAny<Props>;
  };

export interface LinComponentOptions<Data, Props, Methods, Instance> {
  data?: Data;
  field?: boolean;
  classes?: string[];
  mixins?: string[];
  props?: Props & Weapp.PropertyOption;
  relation?: Weapp.RelationOption<Instance> & {
    type: "ancestor" | "descendant";
    name: string;
  };
  relations?: {
    [componentName: string]: Weapp.RelationOption<Instance>;
  };
  methods?: Methods & Weapp.MethodOption<Instance>;

  watch?: Methods & Weapp.MethodOption<Instance>;

  // lifetimes
  beforeCreate?: (this: Instance) => void;
  created?: (this: Instance) => void;
  mounted?: (this: Instance) => void;
  destroyed?: (this: Instance) => void;
}
