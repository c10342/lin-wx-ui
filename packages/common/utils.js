let systemInfo;
export function getSystemInfoSync() {
  if (systemInfo == null) {
    systemInfo = wx.getSystemInfoSync();
  }

  return systemInfo;
}
