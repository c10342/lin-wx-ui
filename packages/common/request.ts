import { isFunction } from "./is";

type RequestOptions = WechatMiniprogram.RequestOption & {
  completeCallback?: Function;
};

// 封装请求方法
const baseRequest = ({
  url,
  data = {},
  header = {},
  timeout,
  method = "GET",
  dataType = "json",
  responseType = "text",
  enableHttp2 = false,
  enableQuic = false,
  enableCache = false,
  completeCallback
}: RequestOptions) => {
  return new Promise((resolve, reject) => {
    wx.request({
      url,
      data,
      header,
      timeout,
      method,
      dataType,
      responseType,
      enableHttp2,
      enableQuic,
      enableCache,
      success(res) {
        if (res.statusCode === 200 && res.errMsg === "request:ok") {
          resolve(res.data);
        } else {
          reject(res);
        }
      },
      // reject,
      complete(resData) {
        if (isFunction(completeCallback)) {
          completeCallback(resData);
        }
      }
    });
  });
};

const request = {
  get(params: RequestOptions) {
    return baseRequest({
      ...params,
      method: "GET"
    });
  },
  post(params: RequestOptions) {
    return baseRequest({
      ...params,
      method: "POST"
    });
  }
};

export default request;
