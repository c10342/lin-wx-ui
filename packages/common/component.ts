import { LinComponentOptions, CombinedComponentInstance } from "./types/index";

const relationFunctions = {
  ancestor: {
    linked(parent) {
      // @ts-ignore
      this.parent = parent;
    },
    unlinked() {
      // @ts-ignore
      this.parent = null;
    }
  },
  descendant: {
    linked(child) {
      // @ts-ignore
      this.children = this.children || [];
      // @ts-ignore
      this.children.push(child);
    },
    unlinked(child) {
      // @ts-ignore
      this.children = (this.children || []).filter((it) => it !== child);
    }
  }
};

function mapKeys(source: object, target: object, map: object) {
  Object.keys(map).forEach((key) => {
    if (source[key]) {
      target[map[key]] = source[key];
    }
  });
}

function makeRelation(options, LinOptions, relation) {
  const { type, name, linked, unlinked, linkChanged } = relation;
  const { beforeCreate, destroyed } = LinOptions;
  if (type === "descendant") {
    options.created = function () {
      beforeCreate && beforeCreate.bind(this)();
      this.children = this.children || [];
    };
    options.detached = function () {
      this.children = [];
      destroyed && destroyed.bind(this)();
    };
  }
  options.relations = Object.assign(options.relations || {}, {
    [`../${name}/index`]: {
      type,
      linked(node) {
        relationFunctions[type].linked.bind(this)(node);
        linked && linked.bind(this)(node);
      },
      linkChanged(node) {
        linkChanged && linkChanged.bind(this)(node);
      },
      unlinked(node) {
        relationFunctions[type].unlinked.bind(this)(node);
        unlinked && unlinked.bind(this)(node);
      }
    }
  });
}

function LinComponent<Data, Props, Methods>(
  LinOptions: LinComponentOptions<
    Data,
    Props,
    Methods,
    CombinedComponentInstance<Data, Props, Methods>
  > = {}
): void {
  const options: any = {};

  mapKeys(LinOptions, options, {
    data: "data",
    props: "properties",
    mixins: "behaviors",
    methods: "methods",
    beforeCreate: "created",
    created: "attached",
    mounted: "ready",
    relations: "relations",
    destroyed: "detached",
    classes: "externalClasses",
    watch: "observers"
  });

  const { relation } = LinOptions;
  if (relation) {
    makeRelation(options, LinOptions, relation);
  }

  // add default externalClasses
  options.externalClasses = options.externalClasses || [];
  options.externalClasses.push("custom-class");

  // map field to form-field behavior
  if (LinOptions.field) {
    options.behaviors = options.behaviors || [];
    options.behaviors.push("wx://form-field");
  }

  if (options.properties) {
    Object.keys(options.properties).forEach((name) => {
      if (Array.isArray(options.properties[name])) {
        // miniprogram do not allow multi type
        options.properties[name] = null;
      }
    });
  }

  // add default options
  options.options = {
    multipleSlots: true,
    addGlobalClass: true
  };

  Component(options);
}

export { LinComponent };
