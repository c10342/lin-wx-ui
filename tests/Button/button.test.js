import {
  getCompId,
  getElement,
  render,
  querySelector,
  loadTemplate,
} from '../utils';
import { canIUseFormFieldButton } from '../../packages/common/version';

jest.mock('../../packages/common/version');

canIUseFormFieldButton.mockReturnValue(true);

let id;

beforeAll(() => {
  // id要唯一，不能每个测试用例都生成一次，不然会报错
  // 或者可以调用  jest.resetModules()，就可以重复生成
  id = getCompId('Button');
});

describe('属性', () => {
  let comp;
  let button;

  beforeEach(() => {
    comp = render(id);
    button = getElement(comp, '.lin-button');
  });

  // 用于 form 组件，可选值为`submit` `reset`，点击分别会触发 form 组件的 submit/reset 事件
  test('formType', () => {
    // 默认情况
    expect(button.getAttribute('formType')).toBe('');
    // 可选值
    const formTypeList = ['submit', 'reset'];
    for (let i = 0; i < formTypeList.length; i++) {
      const formType = formTypeList[i];
      comp.setData({ formType });
      expect(button.getAttribute('formType')).toBe(formType);
    }
  });

  // 是否禁用按钮
  test('disabled', () => {
    const disabledName = 'lin-button-disabled';
    // 默认情况
    expect(button.hasClassName(disabledName)).toBeFalsy();
    // true的情况
    comp.setData({ disabled: true });
    expect(button.hasClassName(disabledName)).toBeTruthy();
  });

  // 是否为块级元素
  test('block', () => {
    const bolckName = 'lin-button-block';
    // 默认情况
    expect(button.hasClassName(bolckName)).toBeFalsy();
    comp.setData({ block: true });
    expect(button.hasClassName(bolckName)).toBeTruthy();
  });

  // 按钮类型
  test('type', () => {
    // 默认情况
    expect(button.hasClassName('lin-button-default')).toBeTruthy();
    // 可选值
    const typeList = [
      'primary',
      'success',
      'info',
      'warning',
      'danger',
      'default',
      'success',
    ];
    for (let i = 0; i < typeList.length; i++) {
      const type = typeList[i];
      comp.setData({ type });
      expect(button.hasClassName(`lin-button-${type}`)).toBeTruthy();
    }
  });

  // 是否为朴素按钮
  test('plain', () => {
    const plainName = 'lin-button-plain';
    // 默认情况
    expect(button.hasClassName(plainName)).toBeFalsy();
    comp.setData({ plain: true });
    expect(button.hasClassName(plainName)).toBeTruthy();
  });

  // 是否为圆角按钮
  test('round', () => {
    const roundName = 'lin-button-round';
    // 默认情况
    expect(button.hasClassName(roundName)).toBeFalsy();
    comp.setData({ round: true });
    expect(button.hasClassName(roundName)).toBeTruthy();
  });

  // 是否为圆形按钮
  test('circle', () => {
    const roundName = 'lin-button-circle';
    // 默认情况
    expect(button.hasClassName(roundName)).toBeFalsy();
    comp.setData({ circle: true });
    expect(button.hasClassName(roundName)).toBeTruthy();
  });

  // 左侧图标相关属性
  test('icon,iconSize', () => {
    // 默认情况
    const icon = getElement(comp, '.lin-button-icon');
    expect(icon.exists()).toBeFalsy();
    comp.setData({
      icon: 'loading',
      iconSize: '14px',
    });
    expect(icon.exists()).toBeTruthy();
    expect(icon.getAttribute('size')).toBe('14px');
  });

  // 按钮尺寸
  test('size', () => {
    // 默认情况
    expect(button.hasClassName('lin-button-size-default')).toBeTruthy();
    const sizeList = ['default', 'medium', 'small'];
    for (let i = 0; i < sizeList.length; i++) {
      const size = sizeList[i];
      comp.setData({ size });
      expect(button.hasClassName(`lin-button-size-${size}`)).toBeTruthy();
    }
  });

  // loading相关属性
  test('loading,loadingColor,loadingSize', () => {
    const loading = getElement(comp, '.lin-button-loading');
    // 默认情况
    expect(loading.exists()).toBeFalsy();
    const loadingSize = '20px';
    const loadingColor = '#fff';
    comp.setData({
      loading: true,
      loadingSize,
      loadingColor,
    });
    expect(loading.exists()).toBeTruthy();
    expect(loading.getAttribute('color')).toBe(loadingColor);
    expect(loading.getAttribute('size')).toBe(loadingSize);
  });

  // 按钮 dataset，open-type 为 `share` 时，可在 onShareAppMessage 事件的 `event.target.dataset.detail` 中看到传入的值
  test('dataset', () => {
    const dataset = { name: 'zhangsan', age: 12 };
    // 默认情况
    expect(button.getAttribute('data-detail')).toBeNull();
    comp.setData({
      dataset,
    });
    expect(button.getAttribute('data-detail')).toEqual(dataset);
  });

  // 按钮颜色，支持传入 linear-gradient 渐变色
  test('color', () => {
    // 默认情况
    expect(comp.data.baseStyle).toBe('');
    // 非朴素按钮情况
    comp.setData({
      color: 'red',
    });
    expect(button.getAttribute('style').replace(/\s+/g, '')).toBe(
      'color:white;background:red;border-color:red;'
    );
    comp.setData({
      color: 'green',
      plain: true,
    });
    expect(button.getAttribute('style').replace(/\s+/g, '')).toBe(
      'color:green;border-color:green;'
    );
    comp.setData({
      color: 'linear-gradient(to right, #4bb0ff, #6149f6)',
    });
    expect(button.getAttribute('style').replace(/\s+/g, '')).toBe(
      'color:linear-gradient(to right, #4bb0ff, #6149f6);border:none;'.replace(
        /\s+/g,
        ''
      )
    );
  });
});

describe('事件', () => {
  let comp;
  let button;
  beforeEach(() => {
    comp = render(id);
    button = getElement(comp, '.lin-button');
  });
  // 点击按钮
  test('click', async () => {
    let fn;
    // 默认情况
    fn = await button.dispatchEvent('tap');
    expect(fn).toBeCalled();
    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).toBeCalledWith('click');

    // 禁用情况
    comp.setData({
      disabled: true,
    });
    fn = await button.dispatchEvent('tap');
    expect(fn).not.toBeCalled();
  });
});

describe('插槽', () => {
  test('default', () => {
    const comp = loadTemplate({
      usingComponents: {
        'lin-button': id,
      },
      template: `<lin-button class='lin-button'>默认按钮</lin-button>`,
    });
    const element = querySelector(comp, '.lin-button');
    expect(element.toJSON()).toMatchSnapshot();
  });
});

describe('外部样式类', () => {
  test('externalClasses', () => {
    const comp = loadTemplate({
      usingComponents: {
        'lin-button': id,
      },
      template: `
        <lin-button 
        custom-class='class-custom'
        loading-class='class-loading'
        icon-class='class-icon'
        hover-class='class-hover'
        class='lin-button'
        ></lin-button>
        `,
    });
    const element = getElement(comp, '.lin-button');
    expect(
      element.hasExternalClass('custom-class', 'class-custom')
    ).toBeTruthy();
    expect(
      element.hasExternalClass('loading-class', 'class-loading')
    ).toBeTruthy();
    expect(element.hasExternalClass('icon-class', 'class-icon')).toBeTruthy();
    expect(element.hasExternalClass('hover-class', 'class-hover')).toBeTruthy();
  });
});
