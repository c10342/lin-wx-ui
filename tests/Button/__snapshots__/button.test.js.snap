// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`插槽 default 1`] = `
<main
  class="lin-button"
>
  <wx-button
    appParameter=""
    ariaLabel=""
    businessId="{{0}}"
    class="custom-class lin-button lin-button-default      lin-button-size-default"
    data-detail="{{null}}"
    formType=""
    hoverClass="hover-class lin-button-hover"
    id=""
    lang=""
    openType=""
    sendMessageImg=""
    sendMessagePath=""
    sendMessageTitle=""
    sessionFrom=""
    showMessageCard="{{false}}"
    style=""
    bind:contact="bindContact"
    bind:error="bindError"
    bind:getphonenumber="bindGetPhoneNumber"
    bind:getuserinfo="bindGetUserInfo"
    bind:launchapp="bindLaunchApp"
    bind:opensetting="bindOpenSetting"
    bind:tap="onClick"
  >
    默认按钮
  </wx-button>
</main>
`;
